Signature recognition

uxmal opened the pull request.

uxmal commented:
@Cairn23: first of all, thanks for the contribution, especially of such an ambitious new feature. It will be a great addition to Reko.

It is the nature of code reviews that they are critical in tone. This may be discouraging to newcomers to the project, who aren't familiar with the coding style and design of the Reko project. Please take the code review in the spirit it is offered. I encourage you to be curious about the review feedback and participate in discussion to clarify coding and design choices. Ultimately, we're all working to improve the code base.

This is a large PR, with many file modifications. This makes the reviewing of the code changes more challenging. It also means that the review process will take longer, perhaps a couple of days. Please have patience as we work through the files.

When submitted, a PR must be mergeable with the target branch (in this case, `master`). This is currently not the case. I encourage you to do a git pull + merge to get the latest changes of `master` into your `Signature-recognition` branch.

When submitted, a PR must furthermore pass all unit tests and pass all the tests in the regression suite (under `$REKO/subjects`). Currently this is not the case. Once you've fixed the merge conflics, please run all the unit tests and regression tests to make sure no failures have been introduced.

One specific note: a big refactoring of the `ElfLoader.cs` file was done, but with code changes rolled in. This makes it much harder to review the changes. In the future, please try to make refactorings (that don't change functionality) in separate commits from feature/bugfix commits. 

Thanks again, and stay tuned as I work through this PR.

uxmal commented:
I've created a new branch `Signature-recognition` in my repo to which this PR will be merged. I'm expecting work will need to continue on this branch after the PR is accepted; it's not quite ready for `master` yet.

uxmal commented:
There were 90 comments in my preliminary pass over the PR. The ball is in your court now, @Cairn23 :)

ptomin commented:
I do not like idea to change interface of `ProcedureBase`. I think, better way is adding library procedures to `Program.ImageSymbols`. New `ProvenanceType ` field can be added to `ImageSymbol` class

Cairn23 commented on src/tools/makesigs/Signatures.txt:
```diff
@@ -0,0 +1,59 @@
+so then signature analysis is carried out the decompiler can automatically rename procedures tothe libray name.
+This can assit the person doing the investigation to either find a procedure in the exectuable code, of help illiminate
+chucks of code fomr the investigation. Either way its a help.
+
```
Issues addresses

Cairn23 commented on src/Core/Core.csproj:
```diff
@@ -736,6 +736,7 @@
     <Compile Include="Services\ITypeLibraryLoaderService.cs" />
     <Compile Include="Services\IUnpackerService.cs" />
     <Compile Include="SignatureBuilder.cs" />
+    <Compile Include="SignitureEntry.cs" />
```
Renamed

Cairn23 commented on src/Core/ProcedureBase.cs:
```diff
@@ -27,24 +27,70 @@
 
 namespace Reko.Core
 {
+    public enum NameStatus
```
added ByteSignature to ProvenanceType and then updated code to reflect the change

Cairn23 commented on src/Core/ProcedureBase.cs:
```diff
@@ -27,24 +27,70 @@
 		public ProcedureBase(string name)
 		{
-			this.Name = name;
+            nameAssignedBy = NameStatus.SystemAssigned;
```
Done

Cairn23 commented on src/Core/ProcedureBase.cs:
```diff
@@ -27,24 +27,70 @@
         private string name;
+        private NameStatus nameAssignedBy;
+
+        public void UpdateNameWithSignatureName(string sigName)
```
Done

Cairn23 commented on src/Core/ProcedureBase.cs:
```diff
@@ -27,24 +27,70 @@
+            NameChanged.Fire(this);
+        }
+
+        public NameStatus NameAssignedBy
```
Done

Cairn23 commented on src/Core/SignitureEntry.cs:
```diff
@@ -0,0 +1,39 @@
+#region License
```
Done

Cairn23 commented on src/Core/SignitureEntry.cs:
```diff
@@ -0,0 +1,39 @@
+        public string Name;
+        public int Length;
+        public byte[] Data;
+        public List<uint> MissBytes;
```
Comment added

Cairn23 commented on src/Decompiler/Analysis/UserSignatureBuilder.cs:
```diff
@@ -73,7 +73,7 @@ public bool BuildSignature(Address addr, Procedure proc)
                     if (sig != null)
                     {
-                        proc.Name = sProc.Name;
+                        proc.UpdateNameByUser(sProc.Name);
```
Done

Cairn23 commented on src/Decompiler/Decompiler.cs:
```diff
@@ -92,6 +96,7 @@ public void Decompile(string filename, string loader = null)
                 Load(filename, loader);
                 ExtractResources();
                 ScanPrograms();
+                ApplySignitures();
```
Done

Cairn23 commented on src/Decompiler/Decompiler.cs:
```diff
@@ -132,6 +137,162 @@ public virtual void AnalyzeDataFlow()
+        {
+            foreach (Program program in Project.Programs)
+            {
+                ApplySigniture(program);
```
Done

Cairn23 commented on src/Decompiler/Decompiler.cs:
```diff
@@ -132,6 +137,162 @@ public virtual void AnalyzeDataFlow()
+            }
+        }
+
+        public virtual void ApplySigniture(Program program)
```
Done

Cairn23 commented on src/Decompiler/Decompiler.cs:
```diff
@@ -132,6 +137,162 @@ public virtual void AnalyzeDataFlow()
+                var eventListener = services.RequireService<DecompilerEventListener>();
+                eventListener.ShowStatus("Performing signature analysis.");
+
+                // Get the architeture for the program so we know which directory of sig files to load
```
Disagree as the eventListener is used within the method to provide feedback, no change

Cairn23 commented on src/Decompiler/Decompiler.cs:
```diff
@@ -132,6 +137,162 @@ public virtual void AnalyzeDataFlow()
+
+                // Get the architeture for the program so we know which directory of sig files to load
+                string name = program.Architecture.Name;
+                string path = Directory.GetCurrentDirectory();
```
Done, but I personally hate the use of var, it's lazy type naming

Cairn23 commented on src/Decompiler/Decompiler.cs:
```diff
@@ -132,6 +137,162 @@ public virtual void AnalyzeDataFlow()
+
+                foreach (string fileName in fileEntries)
+                {
+                    signitureEngine.LoadSignitures(fileName);
```
Done

Cairn23 commented on src/Decompiler/Decompiler.cs:
```diff
@@ -132,6 +137,162 @@ public virtual void AnalyzeDataFlow()
+                }
+
+                // Create instance of sig flirt engine
+                SignatureMatcher signitureEngine = new SignatureMatcher(services);
```
Done

Cairn23 commented on src/Decompiler/Decompiler.cs:
```diff
@@ -132,6 +137,162 @@ public virtual void AnalyzeDataFlow()
+                            var linStart = funcStartAddress.ToLinear();
+
+                            // Take the address of the last statement and then add the length of the statement, which will give us the last byte of the method.
+                            ulong linEnd = endBlock.Statements[endBlock.Statements.Count - 1].LinearAddress + (ulong) lastInstLength;
```
What is been carried out here is to ensure the bytes are in the same order than in a library. The  reason for doing this is that the order of blocks is garrentted. With respect to linkers they update the code to deal with address changes, which is why we have the missingbytes bits within the code.

Cairn23 commented on src/Decompiler/Decompiler.cs:
```diff
@@ -540,6 +701,35 @@ private void ScanProgram(Program program)
             }
         }
 
+        public void LibraryIdentifcation()
```
Done

Cairn23 commented on src/Decompiler/Decompiler.cs:
```diff
@@ -540,6 +701,35 @@ private void ScanProgram(Program program)
+            }
+        }
+
+        private void LibraryIdentifcationProgram(Program program)
```
Done

Cairn23 commented on src/Decompiler/Decompiler.cs:
```diff
@@ -540,6 +701,35 @@ private void ScanProgram(Program program)
+        {
+            try
+            {
+                eventListener.ShowStatus("Scanning code for librarys and applying signitures.");
```
Done

Cairn23 commented on src/Decompiler/Decompiler.cs:
```diff
@@ -540,6 +701,35 @@ private void ScanProgram(Program program)
+            }
+            catch (Exception ex)
+            {
+                eventListener.Error(new NullCodeLocation(""), ex, "Error when Identifying and renaming methods from Signature files.");
```
done

Cairn23 commented on src/Decompiler/Decompiler.cs:
```diff
@@ -540,6 +701,35 @@ private void ScanProgram(Program program)
+            {
+                eventListener.ShowStatus("Scanning code for librarys and applying signitures.");
+                ApplySigniture(program);
+                eventListener.ShowStatus("Finished Scanning code for librarys and applying signitures.");
```
done

Cairn23 commented on src/Decompiler/Decompiler.cs:
```diff
@@ -540,6 +701,35 @@ private void ScanProgram(Program program)
+            {
+                eventListener.Error(new NullCodeLocation(""), ex, "Error when Identifying and renaming methods from Signature files.");
+            }
+            finally
```
removed

Cairn23 commented on src/Decompiler/Loading/SignatureMatcher.cs:
```diff
@@ -0,0 +1,184 @@
+        {
+            SignatureMap = new Dictionary<int, List<SigItem>>();
+            this.eventListener = services.RequireService<DecompilerEventListener>();
+        }
```
Added, but I do thing the attribute naming is lazy

uxmal replied:
Are you referring to the name `fsSvc`? It's used all over Reko, so it's internally consistent.

Cairn23 commented on src/Decompiler/Loading/SignatureMatcher.cs:
```diff
@@ -0,0 +1,184 @@
+        /// <returns>True if the signature was loaded correctly</returns>
+        public bool LoadSignitures(string fileName)
+        {
+            if (File.Exists(fileName))
```
Done

Cairn23 commented on src/Decompiler/Loading/SignatureMatcher.cs:
```diff
@@ -0,0 +1,184 @@
+                    string filename = fileName;
+
+                    //Open the file written above and read values from it.
+                    Stream stream = File.Open(fileName, FileMode.Open);
```
Will leave as is, why replace 4 lines of code which are only called from one place with a separate function

Cairn23 commented on src/Decompiler/Loading/SignatureMatcher.cs:
```diff
@@ -0,0 +1,184 @@
+        /// </summary>
+        /// <param name="length">L</param>
+        /// <returns>True if a signature of the specified lenght exists</returns>
+        bool IsThereSignitureOfLength(int length)
```
Done

Cairn23 commented on src/Decompiler/Loading/SignatureMatcher.cs:
```diff
@@ -0,0 +1,184 @@
+
+
+
+        public string FindMatchingSignitureStart(ICodeLocation currentName, byte[] data, int dataLength)
```
Done

Cairn23 commented on src/Decompiler/Loading/SignatureMatcher.cs:
```diff
@@ -0,0 +1,184 @@
+                List<SigItem> lengthList = SignatureMap[dataLength];
+
+                List<SigItem> matches = new List<SigItem>();
+                foreach (SigItem item in lengthList)
```
Performance is this is not slow, over  a complete decompliation process of 30 seconds, this is taking less then 1 second

uxmal replied:
Have you tested this with large numbers of libraries, though? Having just spent a good week struggling to get rid of O(n ^ 2) behaviour elsewhere in the code, I'm sensitive to this. O(n ^ 2) works fine on small samples, but once you start attempting large files, it blows up terribly.

Cairn23 commented on src/Decompiler/Loading/SignatureMatcher.cs:
```diff
@@ -0,0 +1,184 @@
+                    //eventListener.
+                    string message = "The address function has mutiple options form signature analysis, the options are :- ";
+
+                    foreach(SigItem item in matches)
```
I have already in effect done this by placing the info in the diagnostics window, allow the user to review and update themselves

uxmal replied:
Understood. What I'm proposing is to go the extra mile to help the user by providing them with a piece of user interface that does this. The diagnostics window is not the right place to show large amounts of information; better to pop up a dialog box if the user clicks on it. As I mentioned, I'm not requiring this change to pass the PR. just a comment indicating the possibility of improving the user experience.

Cairn23 commented on src/Decompiler/Loading/SignatureMatcher.cs:
```diff
@@ -0,0 +1,184 @@
+                {
+                    for (int index = 0; index < lengthList.Count; index++)
+                    {
+                        for (int index2 = index + 1; index2 < lengthList.Count; index2++)
```
This not used, method removed

Cairn23 commented on src/Decompiler/Scanning/Scanner.cs:
```diff
@@ -496,13 +496,13 @@ public void ScanImageSymbol(ImageSymbol sym, bool isEntryPoint)
                         var loader = Program.CreateTypeLibraryDeserializer();
                         var exp = loader.LoadExternalProcedure(sProc);
-                        proc.Name = exp.Name;
+                        proc.UpdateNameBySystem(exp.Name);
```
Done

Cairn23 commented on src/Drivers/CmdLine/App.config:
```diff
@@ -1,28 +1,39 @@
+    </switches>
   </system.diagnostics>
-</configuration>
+  <system.web>
```
Don't know why this has been added, but removed

Cairn23 commented on src/Environments/Windows/win32characteristics.xml:
```diff
@@ -29,7 +29,7 @@
   <entry name="ExitProcess">
     <characteristics>
-      <terminates>true</terminates>
+      <terminates>false</terminates>
```
This required to ensure a complete match is made, no change

uxmal replied:
This needs to be changed to `true`; otherwise you will be breaking the parts of Reko's data flow analysis that depend on terminating functions being reported as such. I would suggest that you create a unit test to indicate where this breaks the "complete match" you mentioned in your comment.

Cairn23 commented on src/Environments/Windows/win64characteristics.xml:
```diff
@@ -12,7 +12,7 @@
   <entry name="ExitProcess">
     <characteristics>
-      <terminates>true</terminates>
+      <terminates>false</terminates>
```
This required to ensure a complete match is made, no change

Cairn23 commented on src/Gui/Commands/Cmd_EditSignature.cs:
```diff
@@ -60,7 +60,7 @@ public override void DoIt()
                     program.User.Procedures[address] = sProc;
                     if (procedure != null)
-                        procedure.Name = sProc.Name;
+                        procedure.UpdateNameByUser(sProc.Name);
```
Done

Cairn23 commented on src/Gui/Forms/LibraryIdentifcationPageInteractor.cs:
```diff
@@ -0,0 +1,160 @@
+#region License
```
No action carried out. The reason for doing it this way allows the scope for a skip over analysis if we decide to add it, hence why this has been added

uxmal replied:
I'm not understanding your comment. As mentioned, we should probably discuss this on gitter to hammer out the [user stories](https://en.wikipedia.org/wiki/User_story) relevant to byte signature matching.

Cairn23 commented on src/Gui/Forms/MainFormInteractor.cs:
```diff
@@ -60,6 +60,7 @@ public class MainFormInteractor :
         private IPhasePageInteractor currentPhase;
         private InitialPageInteractor pageInitial;
         private IScannedPageInteractor pageScanned;
+        private ILibraryIdentifcationInteractor pageLibIdent;
```
See above

Cairn23 commented on src/ImageLoaders/Ar/ArHeader.cs:
```diff
@@ -0,0 +1,60 @@
+
+namespace Reko.ImageLoaders.Ar
+{
+    public class ArHeader
```
See ArLoader

Cairn23 commented on src/ImageLoaders/Ar/ArHeader.cs:
```diff
@@ -0,0 +1,60 @@
+            var header = new ArHeader();
+            byte[] data = new byte[60];
+
+            Buffer.BlockCopy(rdr.Bytes, (int) rdr.Offset, data, 0, ArHeader.Size);
```
The rdr does not provide the functionality as required, hence the implementation, nothing updated

uxmal replied:
What is the missing functionality? Could it be added to `ImageReader`?

Cairn23 commented on src/ImageLoaders/Ar/ArLoader.cs:
```diff
@@ -0,0 +1,90 @@
+{
+    public class ArLoader : ImageLoader
+    {
+        // https://en.wikipedia.org/wiki/Ar_(Unix)
```
nothing to do

Cairn23 commented on src/ImageLoaders/Ar/ArLoader.cs:
```diff
@@ -0,0 +1,90 @@
+
+namespace Reko.ImageLoaders.Ar
+{
+    public class ArLoader : ImageLoader
```
Done

Cairn23 commented on src/ImageLoaders/Ar/ArLoader.cs:
```diff
@@ -0,0 +1,90 @@
+            }
+        }
+
+        public int NumberArPackages
```
Nothing done

Cairn23 commented on src/ImageLoaders/Ar/ArPackage.cs:
```diff
@@ -0,0 +1,171 @@
+     
+            // Member name
+            Name = package.header.Name;
+            if (String.Compare(Name, 0, "// ", 0, 3, true) == 0)
```
Done

Cairn23 commented on src/ImageLoaders/Ar/DataTypes.cs:
```diff
@@ -0,0 +1,128 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
```
Done

Cairn23 commented on src/ImageLoaders/Ar/DataTypes.cs:
```diff
@@ -0,0 +1,128 @@
+        public static ArFileType GetFileType(byte[] data)
+        {
+            uint t = GetUint32(data, 0);
+            if (GetUint32(data, 0) == ELF_MAGIC)
```
There is no difference in terms of functionality, given it is already working leaving, if it an't broke don't fix it!

Cairn23 commented on src/ImageLoaders/Ar/DataTypes.cs:
```diff
@@ -0,0 +1,128 @@
+            } 
+            else if (GetUint32(data, 0) == 0xFFFF0000 || GetUint32(data, 0) == 0x10000)
+            {
+                
```
removed

Cairn23 commented on src/ImageLoaders/Ar/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,36 @@
+﻿using System.Reflection;
```
Done

Cairn23 commented on src/ImageLoaders/Coff/CoffLoader.cs:
```diff
@@ -18,24 +18,85 @@
+        //SCOFF_IMAGE_DATA_DIRECTORY pImageDirs;    // Pointer to image directories (for executable files)
+        uint NumImageDirs;                          // Number of image directories (for executable files)
+        uint EntryPoint;                            // Entry point (for executable files)
+        List<SignatureEntry> signitures;
```
Done

Cairn23 commented on src/ImageLoaders/Coff/CoffLoader.cs:
```diff
@@ -49,33 +110,291 @@ public override Program Load(Address addrLoad)
+            {
+                // File has DOS stub
+                rdr.Offset = 0;
+                uint Signature = rdr.ReadUInt32(0x3C);
```
ox3C is used to signify the end of a procedure

Cairn23 commented on src/ImageLoaders/Coff/CoffLoader.cs:
```diff
@@ -49,33 +110,291 @@ public override Program Load(Address addrLoad)
+                if (Signature + 8 < rdr.Bytes.Length && rdr.ReadUInt16((int) Signature) == 0x4550)
+                {
+                    // Executable PE file
+                    FileHeaderOffset = (int) (Signature + 4);
```
comment incorrect and removed

Cairn23 commented on src/ImageLoaders/Coff/CoffLoader.cs:
```diff
@@ -49,33 +110,291 @@ public override Program Load(Address addrLoad)
+            if (StringTableSize > 0)
+            {
+                int startPoint = stringTableOffset + 4;
+                int CharCount = 0;
```
Done

Cairn23 commented on src/ImageLoaders/Coff/OptionHeader.cs:
```diff
@@ -0,0 +1,267 @@
+
+namespace Reko.ImageLoaders.Coff
+{
+
```
It is possible to have this in a COFF file, although options are only used in exes, Its here for completeness of the spec

Cairn23 commented on src/ImageLoaders/Elf/ElfLoader.cs:
```diff
@@ -114,6 +117,10 @@ protected ElfLoader()
         public Dictionary<long, ElfDynamicEntry> DynamicEntries { get; private set; }
         public List<string> Dependencies { get; private set; }
 
+        private List<string> FunctionNames;
```
Moved

Cairn23 commented on src/ImageLoaders/Elf/ElfLoader.cs:
```diff
@@ -741,916 +756,94 @@ protected string rwx(uint flags)
-        public override bool IsExecutableFile { get { return Header64.e_type != ElfImageLoader.ET_REL; } }
 
-        public override ulong AddressToFileOffset(ulong addr)
+        byte[] GetSectionData(ElfSection section)
```
No, cannot use SegmentMap, I did look at this but what we are doing here is getting the library byte codes

Cairn23 commented on src/Reko-decompiler.sln:
```diff
@@ -424,6 +424,10 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "zSeries", "Arch\zSeries\zSe
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "MSP430", "Arch\MSP430\MSP430.csproj", "{FF9E9894-F7E2-464D-B288-4157B864D053}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "RekoSig", "tools\RekoSig\RekoSig.csproj", "{1BEE788B-FF62-4C55-8AFB-CDEAE17B6B3B}"
```
Done

Cairn23 commented on src/tools/makesigs/Macho.cs:
```diff
@@ -0,0 +1,283 @@
+{
+    internal class Macho : DataBuffer
+    {
+
```
This is no longer apart of the project

Cairn23 commented on src/ImageLoaders/Coff/SectionHeader.cs:
```diff
@@ -18,24 +18,175 @@
+                        case Relocation.COFF32_RELOC_REL32:  // 32-bit EIP-relative
+                        case Relocation.COFF32_RELOC_SECREL:   // 32-bit section-relative
+                            {
+                                exemptList.Add(reco.VirtualAddress - start);
```
Not sure if this is correct for an IntervalTree, what is happening here is building a list of all the bytes to skip in the matching process

Cairn23 commented on src/ImageLoaders/Elf/ElfLoader.cs:
```diff
@@ -741,916 +756,94 @@ protected string rwx(uint flags)
 
-        public override List<string> GetDependencyList(byte[] rawImage)
+        
+        public void GenarateSignatures()
```
Nope

Cairn23 commented on src/UnitTests/ImageLoaders/Coff/CoffHeaderTest.cs:
```diff
@@ -0,0 +1,28 @@
+using NUnit.Framework;
```
Added

Cairn23 commented on src/UnitTests/ImageLoaders/Coff/CoffHeaderTest.cs:
```diff
@@ -0,0 +1,28 @@
+namespace Reko.UnitTests.ImageLoaders.Coff
+{
+    [TestFixture]
+    class CoffHeaderTest
```
yes, but what has been added here is in comparison the same as the other tests to be fair

Cairn23 commented on src/UserInterfaces/WindowsForms/WindowsServiceFactory.cs:
```diff
@@ -109,6 +109,11 @@ public IScannedPageInteractor CreateScannedPageInteractor()
             return new ScannedPageInteractor(services);
         }
 
+        public ILibraryIdentifcationInteractor CreateLibraryPageInteractor()
```
As above

Cairn23 commented on src/UserInterfaces/WindowsForms/Forms/DeclarationFormInteractor.cs:
```diff
@@ -246,7 +246,7 @@ private void ModifyDeclaration()
                     up.CSignature = CSignature;
                 if (proc != null)
-                    proc.Name = procName;
+                    proc.UpdateNameByUser(procName);
```
Done

Cairn23 commented on src/tools/RekoSig/SigItem.cs:
```diff
@@ -0,0 +1,59 @@
+using System;
+using System.Collections.Generic;
+
+namespace RekoSig
```
Done

Cairn23 commented on src/tools/makesigs/Coff.cs:
```diff
@@ -0,0 +1,665 @@
+using System;
```
Code is not apart of the project

Cairn23 commented on src/tools/makesigs/Elf.cs:
```diff
@@ -0,0 +1,364 @@
+using System;
```
Code is not part of the project

Cairn23 commented on src/tools/makesigs/FileBuffer.cs:
```diff
@@ -0,0 +1,151 @@
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
```
This OK as is as it apart of the signature generation. Leaving as is.

Cairn23 commented on src/tools/makesigs/Library.cs:
```diff
@@ -0,0 +1,398 @@
+using Reko.ImageLoaders.Coff;
```
Not apart of the project

Cairn23 commented on src/tools/makesigs/Program.cs:
```diff
@@ -1,15 +1,91 @@
+            SignitureGenerator buf = new SignitureGenerator(services, sourceFile);
+            if (buf.GenerateSigniture(destFile))
+            {
+                Console.WriteLine("Signiture generation completed");
```
Corrected

Cairn23 commented on src/tools/makesigs/Program.cs:
```diff
@@ -1,15 +1,91 @@
+            else
+            {
+                Console.WriteLine("Error generating the signitures");
+            }
```
Corrected

Cairn23 commented on src/tools/makesigs/UNIXLibraryHeader.cs:
```diff
@@ -0,0 +1,56 @@
+#region License
+/* 
```
Not apart of the project

Cairn23 commented:
The conflict is noting to do with the work I have done.

uxmal commented:
> The conflict is noting to do with the work I have done.

You refactored `ElfLoader.cs`, moving pieces of it into two separate files. In the meantime, changes were made to `ElfLoader.cs`. The changes have to be accurately merged into your new `ElfLoader.cs`, `ElfLoader32.cs` and `ElfLoader64.cs`.

Cairn23 commented:
Merge of of Uxmal master with this branch carried out. resolved issues in ELFLoader.cs

uxmal commented:
> I do not like idea to change interface of `ProcedureBase`. I think, better way is adding library procedures to `Program.ImageSymbols`. New `ProvenanceType ` field can be added to `ImageSymbol` class

Pavel can you explain what you don't like about the change. Perhaps there is another way to deal with this?

ceeac commented:
I have not looked at the PR in detail, but I think using a tree structure for matching the signatures is faster and more flexible. See for example how IDA does it (conceptually):
https://www.hex-rays.com/products/ida/tech/flirt/in_depth.shtml

uxmal commented:
> I have not looked at the PR in detail, but I think using a tree structure for matching the signatures is faster and more flexible. See for example how IDA does it (conceptually):
> https://www.hex-rays.com/products/ida/tech/flirt/in_depth.shtml

@ceeac: indeed, using a [trie](https://en.wikipedia.org/wiki/Trie) or a [suffix array](https://en.wikipedia.org/wiki/Suffix_array) to perform the pattern matching is likely to be performant. However, the scope of this PR is larger than the algorithm used to perform the actual pattern matching. It involves incorporation pattern recognition into the Reko flow, user interface changes, new file formats, etc. 

Before this PR can be merged there are some issues left to be resolved. There are some unaddressed review comments (see above). The file format for the signature files needs to be properly documented  More seriously, there are no unit tests or regression tests covering the new code. If I wanted to refactor the code to use efficient pattern matching structures like tries or suffix trees, I wouldn't be aware that I was breaking anything. 

@Cairn23: what can I do to assist in completing the PR?

Cairn23 commented:
You need to understand the concepts before even thing about changing the pattern matching. I have compared this to IDA's Flirt and it is the same speed. Given that that doing a full decomplication of one of my test aps, the signature process is only taking less than a 20th of the overall time, I would not worry. if you attempt to use a trie how are you going to account for wild cards, In the pre work I did with this using an array of signatures I was able to create conditions that would mean other then the first couple of bytes you could completely miss match as every byte would have a wild card option. I did start to look at the unit tests, but found that they did not add any value, the same as all the other loader unit test which were just checking the that Image loader code was working placing the data in to the header, there are nor REAL tests within any of the loaders. If you want the tests, please set the correct standard first, before going at people.

Cairn23 commented:
The best thing to do would be to get it added to the master with a beta warning whilst people use it. It is easy to prevent this working by ensuring the signatures are removed.
